Ws ping reqs bug fix

CorgiMan opened the pull request:
Restructure ping-reqs and put them into a indirectPing -> (reached bool, errs []errors) function
The test is said to be inconclusive if **all** nodes that should ping the target are unreachable.

Add tests for all seven possible ping-req scenarios.

Add a ListenerFunc type to the swim/events.go that turns a function into a object that implements EventListener.

jwolski commented on swim/ping_request_sender.go:
```diff
@@ -112,6 +115,26 @@ func (p *pingRequestSender) MakeCall(ctx json.Context, res *pingResponse) <-chan
 	return errC
 }
 
+func indirectHealthCheck(n *Node, target string, size int, timeout time.Duration) (bool, []error) {
```
call it indirectPing. Let's not introduce a new term when there doesn't need to be.

dansimau commented on swim/node.go:
```diff
@@ -218,7 +218,7 @@ func (n *Node) Incarnation() int64 {
 
 func (n *Node) emit(event interface{}) {
 	for _, listener := range n.listeners {
-		go listener.HandleEvent(event)
```
Have we now just made all events synchronous? What is the impact of this?

CorgiMan replied:
The only listeners that are registered are for stats and for hashring updates.

I think the hashring updates are better of being in sync, because we don't want the hashring to be out of sync with the membership.

jwolski replied:
Does this change have anything to do with the bugfix?

CorgiMan replied:
I'm blocking upon receiving certain events to synchronize the ping reqs in the tests in the right way.

danielheller replied:
I think we should have a clear definition of this contract.  Generally, handling events async seems  safer—it allows the event handler implementation to act freely without concern for slowing down other mission-critical paths.  Of course, that does require careful reasoning about how tightly two systems in question are coupled.  If they do need to be tightly coupled to the point of making  operations in the two subsystems atomic, then I think event handling might not be the right pattern.  If synchronous event handling is the right pattern, that contract had better be extremely clearly documented...

dansimau replied:
> If they do need to be tightly coupled to the point of making operations in the two subsystems atomic, then I think event handling might not be the right pattern. 

+1 Good observation.

CorgiMan replied:
Thanks for your thoughts Dan, you haven't been the only one who has expressed worries about the safety of synchronizing the events and it has made me realize my intuition on the observer pattern has been a bit off.

At this moment, events are used for loosely coupled interactions like statting, and strongly coupled interactions updating the hashring.

I strongly believe the hashring should be synchronized with the membership. I'm all for using a different pattern for membership updates anyway.

For testing it is very useful to synchronize the events like I'm doing here. As Simmo and Nils (I believe) have suggested before, having a separate RegisterSynchronized method can help us here.

Since both sync and async events are used in the code and there is no semantic distinction at this point, I'll make a quick writeup on some possible enhancements on the event system on Monday.

CorgiMan commented:
Hold your horses on the review boys, there'll come some changes.

CorgiMan commented:
Ready for review again @uber/ringpop

dansimau commented on swim/ping_request_test.go:
```diff
@@ -119,3 +123,250 @@ func (s *PingRequestTestSuite) TestTimesOut() {
+
+	sender, helper1, helper2, target := tnodes[0], tnodes[1], tnodes[2], tnodes[3]
+	bootstrapNodes(t, sender, helper1, helper2, target)
+	waitForConvergence(t, time.Second/2, sender, helper1, helper2, target)
```
`500*time.Millisecond`?

CorgiMan replied:
I should have acked this earlier

dansimau commented:
Just a general piece of feedback: I found some of the ping tests took a lot of effort to read and understand. I feel that some more comments explaining what is going on would help the reader figure it out faster.

As a general rule, if it takes some effort to design the right pattern (as I know these ping/timings did), it will take some effort for the reader to comprehend. In these situations, being liberal with code comments  can significantly reduce the amount of time the reader needs to get up to speed with what the code is doing.

dansimau commented:
Oh, forgot to say: other than that, the code looks good! :smile:

jwolski commented on swim/ping_request_sender.go:
```diff
@@ -112,6 +115,29 @@ func (p *pingRequestSender) MakeCall(ctx json.Context, res *pingResponse) <-chan
+// indirectPing is used to check if a target node can be reached indirectly.
+// The indirectPing is performed by sending a specifiable amount of ping
+// requests nodes in n's membership.
+func indirectPing(n *Node, target string, amount int, timeout time.Duration) (reached bool, errs []error) {
```
these return values need not be named: `(reached bool, errs []error)` -> `(bool, []error)`

CorgiMan replied:
I think it increases the readability since one might think the first return value could mean isHealthy.
isHealthy has a different meaning than reached when the indirectPing is inconclusive.

I think the renaming of this function from indirectHealthCheck to indirectPing, has helped the interpretability of return value, so I'm okay with removing it. 

I'd still like your thoughts on why we benefit from removing the names though.

jwolski replied:
Only reason I suggested removing them is because of the way I've seen Go's named return values previously used. They're used as a replacement actually returning values in various parts of a method and not to serve as documentation.

CorgiMan replied:
The go tour suggests otherwise: "These names should be used to document the meaning of the return values." https://tour.golang.org/basics/7

They indeed also add some other functionality, which I use a bit off here. (using errs).

jwolski replied:
the tour.golang statement is a bit ambiguous.. but ok you win this time.

dansimau replied:
Also see:
https://golang.org/doc/effective_go.html#named-results
https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

jwolski replied:
Interesting. Effective Go makes it much more explicit. I really lose now.

danielheller commented:
I think the commit message here is a step in the right direction, but could be made more informative/accessible at very little cost in brevity, e.g. "Ping Reqs are defined to fail only if all N requests which are fanned out fail; this change addresses a bug which makes us treat the ping req as failed if *any* of the requests fail."

danielheller commented on swim/ping_request_sender.go:
```diff
@@ -112,6 +115,29 @@ func (p *pingRequestSender) MakeCall(ctx json.Context, res *pingResponse) <-chan
+			// If the ping to the target was not-ok we want to wait for more results.
+
+		case error:
+			errs = append(errs, res)
```
Should we have default case with a panic here, or perhaps use an anonymous struct that we treat as a union?

CorgiMan replied:
I'm a fan of having a struct as a union here. I wanted to refactor only the things I had to in this diff. I'll be happy to refactor it in another diff, please let me know what you think.

danielheller commented on swim/ping_request_sender.go:
```diff
@@ -82,6 +82,9 @@ func (p *pingRequestSender) SendPingRequest() (*pingResponse, error) {
 	select {
 	case err := <-p.MakeCall(ctx, &res):
+		if err == nil {
+			p.node.memberlist.Update(res.Changes)
```
Is this an unrelated bug that you discovered in the course of this change, i.e. that we aren't applying changes gossiped in ping reqs?  Should this be a separate diff with its own tests?

CorgiMan replied:
This was handled in the pingNextMember before, which worked since thats the only caller of sendPingRequests. Handling it there is an accident waiting to happen if we would ever send ping-reqs for another reason. I believe it's should be part of this diff because it's part of refactoring the ping-reqs mechanism.

CorgiMan commented:
Noted, thanks for the insight. It is not entirely correct though. The bug is that the indirect check has the same result as the first ping-req that comes back (positive or negative). We talked about this and I misspoke then, hence the confusion.

jwolski commented on swim/ping_request_test.go:
```diff
@@ -119,3 +123,250 @@ func (s *PingRequestTestSuite) TestTimesOut() {
+}
+
+// The first ping-req doesn't reach the helper node. The second ping-req
+// does and the ping reached the target. The indirectPing therefor return
```
`therefore`

CorgiMan replied:
ack

jwolski commented on swim/ping_request_test.go:
```diff
@@ -119,3 +123,250 @@ func (s *PingRequestTestSuite) TestTimesOut() {
+	tnodes := genChannelNodes(t, 4)
+	defer destroyNodes(tnodes...)
+
+	// don't boostrap the target
```
`bootstrap`

CorgiMan replied:
ack

jwolski commented on swim/ping_request_test.go:
```diff
@@ -119,3 +123,250 @@ func (s *PingRequestTestSuite) TestTimesOut() {
+	assert.Len(t, errs, 2, "expected only errors from the helper nodes")
+}
+
+// onPingRequestReceive can be registered on an EventListener and fires the
```
`onPingRequestComplete`

CorgiMan replied:
Thanks

jwolski commented on swim/ping_request_test.go:
```diff
@@ -119,3 +123,250 @@ func (s *PingRequestTestSuite) TestTimesOut() {
+
+	bootstrapNodes(t, tnodes...)
+	waitForConvergence(t, time.Second/2, tnodes...)
+	sender, helper1, helper2, target := tnodes[0], tnodes[1], tnodes[2], tnodes[3]
```
Can we make waitForTChannelClosed part of the testNode type?

```
helper1.closeAndWait(sender.tchannel)
```

I think it could help DRY things up a bit.

CorgiMan replied:
Sure, I don't mind repeating myself in tests, but I don't mind changing it either

CorgiMan commented:
@dansimau @jwolski @danielheller  I have addressed your feedback
(Typos, time.Second/2 -> 500*time.Millisecond, and DRYing up with closeAndWait function on testNode)

dansimau commented:
LGTM. Integration tests have landed in ringpop-go since this PR.

Is there a test in there for this fix? Is it passing?

CorgiMan commented:
There is not test for this yet. I'll see if I can add it.

CorgiMan commented:
It's not trivial to implement, let's leave it for another day

jwolski commented:
lgtm
